compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala

```diff
@@ -161,7 +163,14 @@ object GenericSignatures {
 
             // TODO revisit this. Does it align with javac for code that can be expressed in both languages?
             val delimiter = if (builder.charAt(builder.length() - 1) == '>') '.' else '$'
-            builder.append(delimiter).append(sanitizeName(sym.name.asSimpleName))
+
+            @tailrec def appendIfNotDerivedName(name: Name): Unit = name match {
+              case _: DerivedName => //do nothing
```

Comment:
This will generate an invalid signature. The point of this code is to generate a Java generic signature for the benefit of java code compiling against scala code, for example if I have a method that returns a `List[String]`, the Java bytecode signature will just be:
```java
()Lscala/collection/immutable/List;
```
since Java bytecode doesn't know anything about generics. This will work, but it's annoying to use from Java, so for the sole benefit of javac, we also generate a generic signature:
```java
()Lscala/collection/immutable/List<Ljava/lang/String;>;
```
(you can see this signature by running `javap -p` on a .class file)
This lets javac know that we're returning a List of String and not a List of anything else.

Anyway, in #6626, the problematic method is:
```scala
def m(): A[B.type] = ???
```
The generic signature for this method should contain `B$` since that's the class that corresponds to the object `B`. The problem is that the name of `B` is a derived name (the name of an object name is always a derived name based on the name of the corresponding class), so calling `asSimpleName` will lead to a crash. But the fix is easy: we're interested in the name of the class in bytecode, which we can obtain by calling `mangledString`, it turns out that `sanitizeName` already calls `mangledString`, so I think we can just remove `.asSimpleName` from the original name and have it work.

Reply:
ah okay

Reply:
that was . good explanation. Can we document this somewhere?